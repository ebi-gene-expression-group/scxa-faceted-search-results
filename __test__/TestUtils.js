import React from 'react'
import PropTypes from 'prop-types'

const EpisodeCard = ({title}) =>
  <div>
    <p>{title}</p>
  </div>

EpisodeCard.propTypes = {
  title: PropTypes.string.isRequired
}

// Stolen from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random
const getRandomInt = (min, max) => {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min)) + min; //The maximum is exclusive and the minimum is inclusive
}

const vindicators = [
  { group:`Vindicator`, value: `supernova`, label: `Supernova`, disabled: false},
  { group:`Vindicator`, value: `noob-noob`, label: `Noob-Noob`, disabled: false},
  { group:`Vindicator`, value: `vance_maximus`, label: `Vance Maximus`, disabled: false},
  { group:`Vindicator`, value: `alan_rails`, label: `Alan Rails`, disabled: false},
  { group:`Vindicator`, value: `crocubot`, label: `Crocubot`, disabled: false},
  { group:`Vindicator`, value: `million_ants`, label: `Million Ants`, disabled: false},
  { group:`Vindicator`, value: `morty_smith`, label: `Morty Smith`, disabled: false},
  { group:`Vindicator`, value: `rick_sanchez`, label: `Rick Sanchez`, disabled: false},
  { group:`Vindicator`, value: `lady_katana`, label: `Lady Katana`, disabled: false},
  { group:`Vindicator`, value: `calypso`, label: `Calypso`, disabled: false},
  { group:`Vindicator`, value: `diablo_verde`, label: `Diablo Verde`, disabled: false}
]

const episodes = [
  {
    element: {
      title: `Raising Gazorpazorp`
    },
    facets: [
      {
        group: `Planet`,
        value: `gazorpazorp`,
        label: `Gazorpazorp`,
        description: `Planet Tooltip`
      },
      {
        group: `Guest character`,
        value: `gwendolyn`,
        label: `Gwendolyn`,
        description: `Guest character Tooltip`
      },
      {
        group: `Guest character`,
        value: `ma-sha`,
        label: `Ma-Sha`,
        description: `Guest character Tooltip`
      },
      {
        group: `Season`,
        value: `1`,
        label: `Season 1`,
        description: `Season Tooltip`
      }
    ]
  },
  {
    element: {
      title: `The wedding squanchers`
    },
    facets: [
      {
        group: `Planet`,
        value: `squanch`,
        label: `Squanch`,
        description: `Planet Tooltip`
      },
      {
        group: `Guest character`,
        value: `birdperson`,
        label: `Birdperson`,
        description: `Guest character Tooltip`
      },
      {
        group: `Guest character`,
        value: `squanchy`,
        label: `Squanchy`,
        description: `Guest character Tooltip`
      },
      {
        group: `Season`,
        value: `2`,
        label: `Season 2`,
        description: `Season Tooltip`
      }
    ]
  },
  {
    element: {
      title: `The Rickshank redemption`
    },
    facets: [
      {
        group: `Guest character`,
        value: `birdperson`,
        label: `Birdperson`,
        description: `Guest character Tooltip`
      },
      {
        group: `Planet`,
        value: `buttworld`,
        label: `Buttworld`,
        description: `Planet Tooltip`
      },
      {
        group: `Season`,
        value: `3`,
        label: `Season 3`,
        description: `Season Tooltip`
      }
    ]
  },
  {
    element: {
      title: `Ricksy business`
    },
    facets: [
      {
        group: `Guest character`,
        value: `squanchy`,
        label: `Squanchy`,
        description: `Guest character Tooltip`
      },
      {
        group: `Guest character`,
        value: `abradolf_lincler`,
        label: `Abradolf Lincler`,
        description: `Guest character Tooltip`
      },
      {
        group: `Season`,
        value: `1`,
        label: `Season 1`,
        description: `Season Tooltip`
      }
    ]
  },
  {
    element: {
      title: `Close Rick-counters of the Rick kind`
    },
    facets: [
      {
        group: `Guest character`,
        value: `ricktiminus_sancheziminius`,
        label: `Ricktiminus Sancheziminius`,
        description: `Guest character Tooltip`
      },
      {
        group: `Guest character`,
        value: `abradolf_lincler`,
        label: `Abradolf Lincler`,
        description: `Guest character Tooltip`
      },
      {
        group: `Planet`,
        value: `buttworld`,
        label: `Buttworld`,
        description: `Planet Tooltip`
      },
      {
        group: `Season`,
        value: `1`,
        label: `Season 1`,
        description: `Season Tooltip`
      }
    ]
  }
]

export {getRandomInt, vindicators, episodes, EpisodeCard}
