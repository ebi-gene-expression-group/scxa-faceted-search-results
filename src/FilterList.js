import React from 'react'
import PropTypes from 'prop-types'
import {ResultPropTypes} from './ResultPropTypes'

const dynamicSort = (property, sortState) => {
  const sortOrder = sortState ?  -1 : 1
  return function (a, b) {
    const result = (a[property] < b[property]) ? -1 : (a[property] > b[property]) ? 1 : 0
    return result * sortOrder
  }
}

const titleCase = (str) =>  
  str.toLowerCase().split(` `).map((string, index)=>
    index === 0 ?
      string.charAt(0).toLowerCase() + string.substring(1) :
      string.charAt(0).toUpperCase() + string.substring(1)    
    ).join(``)

class FilterList extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      sortTitle: titleCase(Object.keys(this.props.ResultsHeaderClass().titles)
        .filter(title => Object.keys(this.props.filteredResults[0].element).includes(titleCase(title)))[0]),
      sortState: true
    }
    this.constructTableHeader = this.constructTableHeader.bind(this)
    this.sortTable = this.sortTable.bind(this)
  }
  
  constructTableHeader(ResultsHeaderClass){
    const tableTitles = Object.keys(ResultsHeaderClass().titles)
    const tableTitleDivs = Object.values(ResultsHeaderClass().titles)
    const ContainerDiv = ResultsHeaderClass().container

    const TableHeader = () => 
      <ContainerDiv>
        { 
          tableTitles.map((title, index) => {
            const TitleDiv = tableTitleDivs[index]
            return Object.keys(this.props.filteredResults[0].element).includes(titleCase(title)) ?
              titleCase(title) === this.state.sortTitle ?
                this.state.sortState ?
                  <TitleDiv key={title} style={{opacity: 1}}> <p id={`selected`} onClick={this.sortTable}>{title} ▼</p></TitleDiv> :
                  <TitleDiv key={title} style={{opacity: 1}}> <p id={`selected`} onClick={this.sortTable}>{title} ▲</p></TitleDiv> 
                :
                <TitleDiv key={title}> <p id={`title`} onClick={this.sortTable}>{title} ▲</p></TitleDiv> 
              :
              <TitleDiv key={title}> <p id={`title`} onClick={this.sortTable}>{title}</p></TitleDiv> 
          }) 
        }
      </ContainerDiv>

    return TableHeader
  }

  sortTable(event) {
    Object.keys(this.props.filteredResults[0].element).includes(titleCase(event.target.innerText).slice(0, -1)) && 
    this.setState({
      sortTitle: titleCase(event.target.innerText).slice(0, -1), 
      sortState: event.target.innerText.slice(-1) === `▲`
    })
  }

  render() {
    const {filteredResults, resultsMessage, ResultsHeaderClass, ResultElementClass} = this.props
    const filteredElements = filteredResults.map((result) => result.element)
    const TableHeader = this.constructTableHeader(ResultsHeaderClass)
    filteredElements.sort(dynamicSort(this.state.sortTitle, this.state.sortState))

    return (
      <div>
        <h4>{resultsMessage}</h4>
        { filteredElements.length && ResultsHeaderClass && <TableHeader /> }
        { filteredElements.map((element, index) => <div key={index}><ResultElementClass {...element}/></div>) }
      </div>
    )
  }
}

FilterList.propTypes = {
  filteredResults: PropTypes.arrayOf(ResultPropTypes).isRequired,
  resultsMessage: PropTypes.string,
  ResultsHeaderClass: PropTypes.func,
  ResultElementClass: PropTypes.func.isRequired
}

FilterList.defaultProps = {
  resultsMessage: ``
}

export default FilterList
