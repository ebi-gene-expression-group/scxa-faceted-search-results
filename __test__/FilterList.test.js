import React from 'react'
import renderer from 'react-test-renderer'
import Enzyme from 'enzyme'
import { mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { getRandomInt, episodes, ExperimentTableHeader, ExperimentTableCard } from './TestUtils'

import FilterList from '../src/FilterList'

Enzyme.configure({ adapter: new Adapter() })

const props = {
  filteredResults: episodes,
  ResultsHeaderClass: ExperimentTableHeader,
  ResultElementClass: ExperimentTableCard
}

describe(`FilterList`, () => {
  test(`renders as many components of ResultElementClass as filtered results`, () => {
    const wrapper = mount(<FilterList {...props} filteredResults={episodes} />)
    expect(wrapper.find(`#title`)).toHaveLength(Object.keys(ExperimentTableHeader().titles).length-1)
    expect(wrapper.find(`#selected`)).toHaveLength(1)
    expect(wrapper.find(ExperimentTableCard)).toHaveLength(episodes.length)
  })

  test(`highlights the user selected table header`, () => {
    const titleNames = Object.keys(ExperimentTableHeader().titles)
    const wrapper = mount(<FilterList {...props} filteredResults={episodes} />)
    const mockState = (str) => {
      const splitStr = str.toLowerCase().split(' ').map((string, index)=>
        index === 0 ?
          string.charAt(0).toLowerCase() + string.substring(1) :
          string.charAt(0).toUpperCase() + string.substring(1)    
      )
      return splitStr.join('')
    }
    wrapper.setState({sortTitle: mockState(titleNames[getRandomInt(0, titleNames.length)])})
    expect(wrapper.find(`#title`)).toHaveLength(1)
    expect(wrapper.find(`#selected`)).toHaveLength(1)
  })

  test(`matches snapshot`, () => {
    const tree = renderer.create(<FilterList {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
