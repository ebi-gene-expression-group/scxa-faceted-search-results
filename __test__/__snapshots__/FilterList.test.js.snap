// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterList matches snapshot 1`] = `
<div>
  <h4>
    
  </h4>
  <div
    className="sc-htpNat jYhomY"
  >
    <div
      className="sc-bdVaJa edgMnc"
    >
       
      <p
        id="title"
        onClick={[Function]}
      >
        Meow
         ↑
      </p>
    </div>
    <div
      className="sc-bwzfXH laSsVf"
    >
       
      <p
        id="title"
        onClick={[Function]}
      >
        Wow wow
         ↑
      </p>
    </div>
  </div>
  <div>
    <div>
      <p>
        Raising Gazorpazorp
      </p>
    </div>
  </div>
  <div>
    <div>
      <p>
        The wedding squanchers
      </p>
    </div>
  </div>
  <div>
    <div>
      <p>
        The Rickshank redemption
      </p>
    </div>
  </div>
  <div>
    <div>
      <p>
        Ricksy business
      </p>
    </div>
  </div>
  <div>
    <div>
      <p>
        Close Rick-counters of the Rick kind
      </p>
    </div>
  </div>
</div>
`;
